### PR TITLE
Refactor to typescript slice 21

### LINKED ISSUE #778: Refactor to typescript slice 21 (Refinements)
This issue aims to do the last cleanup required to finish the refactoring to typescript.

* Remove all the ToDo comments and fix as required.
* Generate the typing files of the contracts with TypeChain.
* Consider reorganizing the domain in a single module and move somewhere else the types that do not belong to the domain.
* Clear all JSDocs param/return types
* Review for eslint-disable rules that can be removed

### PR DESCRIPTION
**Pull request by amonsosanz**
Closes #778.

### What does this PR does?

This PR does the last cleanup required to finish the refactor to typescript. Mainly it does:

* Remove all the ToDo comments and fix as required.
* Generate the typing files of the contracts with TypeChain.
* Reorganizes the domain in a single module.
* Clear all JSDocs param/return types
* Review for eslint-disable rules that can be removed

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Respect code style and lint
- [ ] Update documentation (if needed)

### CODE REVIEW

**Review by elias-garcia — COMMENTED** [review]
Regarding the script for generate the contract types I'm going to make some general suggestions:

1. Rename `contracts` folder in the root of the project to `abis` as at the end these are not contracts but abis (Application Binary Interface).
2. I would avoid to have nested folders in this `abi` folder as each contract has only one abi.
3. I would rename `erc20.json` to `erc-20.json` as the real name of the standard is ERC-20 so it looks more correct I think. But not sure if this is going to work with TypeChain, if it adds the hyphen to the final result or if it does weird things I'll keep the name as it's now.
4. I would create a folder `src/types` and publish the generated result inside the folder `src/types/contracts`.
5. I would also move the `hermezjs-typings.d.ts` to this new `src/types` folder and rename the file to `hermezjs.d.ts`.
6. I would rename the `generate-contract-typings.sh` script to just `generate-contract-types.sh`. The same in the `scripts` section of the `package.json`.

I would also add a quick explanation in the `README.md` to understand how this works. For example in the `npm i` section. Just a line or two explaining that the types for the contracts are going to be generated in the `postinstallation` step and the output is stored in the `/types/contracts` folder.

**Comment by amonsosanz** [comment]
> Regarding the script for generate the contract types I'm going to make some general suggestions:
> 
> 1. Rename `contracts` folder in the root of the project to `abis` as at the end these are not contracts but abis (Application Binary Interface).

Ok, done, although this is not 100% true since for hermez we clone the repo `hermeznetwork/contracts` with the actual contracts.

> 2. I would avoid to have nested folders in this `abi` folder as each contract has only one abi.

While it's true that for the ERC-20 contract we only have the `erc20.json` ABI, I think we need to keep the folders under `abis` to have a target to clone `hermeznetwork/contracts` repo and later remove it. See the last cmd in the script `rm -rf "abis/hermez"`.

> 3. I would rename `erc20.json` to `erc-20.json` as the real name of the standard is ERC-20 so it looks more correct I think. But not sure if this is going to work with TypeChain, if it adds the hyphen to the final result or if it does weird things I'll keep the name as it's now.

Done. TypeChain works ok, although it removes the hyphen in the compiled types:
image

> 4. I would create a folder `src/types` and publish the generated result inside the folder `src/types/contracts`.

Done.

> 5. I would also move the `hermezjs-typings.d.ts` to this new `src/types` folder and rename the file to `hermezjs.d.ts`.

Done.

> 6. I would rename the `generate-contract-typings.sh` script to just `generate-contract-types.sh`. The same in the `scripts` section of the `package.json`.

Done.

> I would also add a quick explanation in the `README.md` to understand how this works. For example in the `npm i` section. Just a line or two explaining that the types for the contracts are going to be generated in the `postinstallation` step and the output is stored in the `/types/contracts` folder.

Done.

File: src/utils/transactions.ts [hunk 1]
```diff
@@ -34,14 +34,10 @@ function getTransactionAmount(
  * If it's an L1 transaction, it adds the forgeDelay again
  */
 function getTxPendingTime(
-  // ToDo: This undefined and the check below can be removed once the views are migrated to TS
-  coordinatorState: CoordinatorState | undefined,
+  coordinatorState: CoordinatorState,
```
[the remark this answers is not in the document]
Comment (amonsosanz) [reply src/utils/transactions.ts]: Done

**Comment by elias-garcia** [comment]
> Ok, done, although this is not 100% true since for hermez we clone the repo hermeznetwork/contracts with the actual contracts.

Now that you mention it and I thought a little bit about this... We don't need the Hermez contracts or any abi related to Hermez contracts, as they are abstracted with `hermezjs`. If the contracts add a new feature, we will create a wrapper for it inside `hermezjs`, so we can remove Hermez contracts from the `typechain` script!

**Comment by amonsosanz** [comment]
> We don't need the Hermez contracts or any abi related to Hermez contracts

Correct, that's why the Hermez repo clone and the generation of the types are commented in the script.

We [already discussed this](https://0xpolygon.slack.com/archives/C02FS3PHEGH/p1640178024042100) and agreed to keep thes lines in the script to provide an example of how one would interact with the contract through the ABI.

**Comment by elias-garcia** [comment]
> > We don't need the Hermez contracts or any abi related to Hermez contracts
> 
> Correct, that's why the Hermez repo clone and the generation of the types are commented in the script.
> 
> We [already discussed this](https://0xpolygon.slack.com/archives/C02FS3PHEGH/p1640178024042100) and agreed to keep these lines in the script to provide an example of how one would interact with the contract through the ABI.

Okay! I'm seeing that @AlbertoElias approved that. I wouldn't keep it as we are not going to need any contract in the wallet and for me this example is more about how to use `typechain` than the wallet... But at least for us makes sense to name the folder `abis` instead of `contracts`, as we only need abis.

**Comment by amonsosanz** [comment]
> > > We don't need the Hermez contracts or any abi related to Hermez contracts
> > 
> > 
> > Correct, that's why the Hermez repo clone and the generation of the types are commented in the script.
> > We [already discussed this](https://0xpolygon.slack.com/archives/C02FS3PHEGH/p1640178024042100) and agreed to keep these lines in the script to provide an example of how one would interact with the contract through the ABI.
> 
> Okay! I'm seeing that @AlbertoElias approved that. I wouldn't keep it as we are not going to need any contract in the wallet and for me this example is more about how to use `typechain` than the wallet... But at least for us makes sense to name the folder `abis` instead of `contracts`, as we only need abis. Also, I would keep this folder without any subdirectories like `abis/erc20/erc20.json`.

As I mentioned before we could remove the folder for the `erc20.json` ABI but not the folder for the hermez contracts, since we need to have a target folder to clone the repo `hermeznetwork/contracts` and later remove it. Maybe I'm missing something?

Then my point is that since we need a folder for hermez, I'd keep the folder for the erc20 ABI, otherwise they are at different levels. Does it make sense to you?

**Review by AlbertoElias — APPROVED** [review]

**Pull request merged**